File: Support/Templates/Files/SwitchableObject.cs
using System;
using Csla;

namespace Templates
{
  [Serializable]
  public class SwitchableObject : BusinessBase<SwitchableObject>
  {
    #region Business Methods

    // TODO: add your own fields, properties and methods

    // example with private backing field
    private static PropertyInfo<int> IdProperty =
      RegisterProperty(new PropertyInfo<int>("Id"));
    private int _Id = IdProperty.DefaultValue;
    public int Id
    {
      get { return GetProperty(IdProperty, _Id); }
      set { SetProperty(IdProperty, ref _Id, value); }
    }

    // example with managed backing field
    private static PropertyInfo<string> NameProperty =
      RegisterProperty(new PropertyInfo<string>("Name"));
    public string Name
    {
      get { return GetProperty(NameProperty); }
      set { SetProperty(NameProperty, value); }
    }

    #endregion

    #region Validation Rules

    protected override void AddBusinessRules()
    {
      // TODO: add validation rules
      //ValidationRules.AddRule(RuleMethod, "");
    }

    #endregion

    #region Authorization Rules

    protected override void AddAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowWrite("Name", "Role");
    }

    private static void AddObjectAuthorizationRules()
    {
      // TODO: add authorization rules
      //AuthorizationRules.AllowEdit(typeof(SwitchableObject), "Role");
    }

    #endregion

    #region Root Factory Methods

    public static SwitchableObject NewSwitchableObject()
    {
      return DataPortal.Create<SwitchableObject>();
    }

    public static SwitchableObject GetSwitchableObject(int id)
    {
      return DataPortal.Fetch<SwitchableObject>(
        new SingleCriteria<SwitchableObject, int>(id));
    }

    public static void DeleteSwitchableObject(int id)
    {
      DataPortal.Delete<SwitchableObject>(new SingleCriteria<SwitchableObject, int>(id));
    }

    #endregion

    #region Child Factory Methods

    internal static SwitchableObject NewSwitchableChild()
    {
      return DataPortal.CreateChild<SwitchableObject>();
    }

    internal static SwitchableObject GetSwitchableChild(object childData)
    {
      return DataPortal.FetchChild<SwitchableObject>(childData);
    }

    private SwitchableObject()
    { /* Require use of factory methods */ }

    #endregion

    #region Root Data Access

    [RunLocal]
    protected override void DataPortal_Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      base.DataPortal_Create();
    }

    private void DataPortal_Fetch(SingleCriteria<SwitchableObject, int> criteria)
    {
      // TODO: load values
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_Insert()
    {
      // TODO: insert values
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_Update()
    {
      // TODO: update values
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    protected override void DataPortal_DeleteSelf()
    {
      DataPortal_Delete(new SingleCriteria<SwitchableObject, int>(this.Id));
    }

    [Transactional(TransactionalTypes.TransactionScope)]
    private void DataPortal_Delete(SingleCriteria<SwitchableObject, int> criteria)
    {
      // TODO: delete values
    }

    #endregion

    #region Child Data Access

    protected override void Child_Create()
    {
      // TODO: load default values
      // omit this override if you have no defaults to set
      base.Child_Create();
    }

    private void Child_Fetch(object childData)
    {
      // TODO: load values
    }

    private void Child_Insert(object parent)
    {
      // TODO: insert values
    }

    private void Child_Update(object parent)
    {
      // TODO: update values
    }

    private void Child_DeleteSelf(object parent)
    {
      // TODO: delete values
    }

    #endregion
  }
}

File: Support/Templates/vb/Files/SwitchableObject.vb
Imports System
Imports Csla
Imports Csla.Security

<Serializable()> _
Public Class SwitchableObject
  Inherits BusinessBase(Of SwitchableObject)

  Public Shared ReadOnly IdProperty As PropertyInfo(Of Integer) = RegisterProperty(Of Integer)(NameOf(Id))
  Public Property Id() As Integer
    Get
      Return GetProperty(IdProperty)
    End Get
    Set(ByVal value As Integer)
      SetProperty(IdProperty, value)
    End Set
  End Property

  Public Shared ReadOnly NameProperty As PropertyInfo(Of String) = RegisterProperty(Of String)(NameOf(Name))
  Public Property Name() As String
    Get
      Return GetProperty(NameProperty)
    End Get
    Set(ByVal value As String)
      SetProperty(NameProperty, value)
    End Set
  End Property

  Protected Overrides Sub AddBusinessRules()
    'call base class implementation to add data annotation rules to BusinessRules 
    MyBase.AddBusinessRules()

    ' TODO: add validation rules
    'BusinessRules.AddRule(New MyRule, IdProperty)

  End Sub

  Public Shared Sub AddObjectAuthorizationRules()
    'TODO: add authorization rules
    'BusinessRules.AddRule(...)
  End Sub

  <Create>
  <RunLocal()>
  Private Sub Create()
    ' TODO: load default values 
    ' omit this override if you have no defaults to set 
    BusinessRules.CheckRules()
  End Sub

  <Fetch>
  Private Sub Fetch(ByVal id As Integer)
    ' TODO: load values 
  End Sub

  <Insert>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub Insert()
    ' TODO: insert values 
  End Sub

  <Update>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub Update()
    ' TODO: update values 
  End Sub

  <DeleteSelf>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Sub DeleteSelf()
    Delete(Me.Id)
  End Sub

  <Delete>
  <Transactional(TransactionalTypes.TransactionScope)>
  Private Shadows Sub Delete(ByVal id As Integer)
    ' TODO: delete values 
  End Sub

  Private Sub CreateChild()
    ' TODO: load default values 
    ' omit this override if you have no defaults to set 
    BusinessRules.CheckRules()
  End Sub

  Private Sub FetchChild(ByVal childData As Object)
    ' TODO: load values 
  End Sub

  Private Sub InsertChild(ByVal parent As Object)
    ' TODO: insert values 
  End Sub

  Private Sub UpdateChild(ByVal parent As Object)
    ' TODO: update values 
  End Sub

  Private Sub DeleteSelfChild(ByVal parent As Object)
    ' TODO: delete values 
  End Sub

End Class
